chainerx_cc/chainerx/native/native_device/reduction.cc

```diff
@@ -76,6 +76,31 @@ class NativeSumKernel : public SumKernel {
 
 CHAINERX_NATIVE_REGISTER_KERNEL(SumKernel, NativeSumKernel);
 
+class NativeProdKernel : public ProdKernel {
+public:
+    void Call(const Array& a, const Axes& axis, const Array& out) override {
+        CHAINERX_ASSERT(internal::IsValidReductionShape(a.shape(), axis, out.shape(), true));
+        a.device().CheckDevicesCompatible(a, out);
+
+        auto do_prod = [&a, &axis, &out](auto in_pt, auto out_pt) {
+            using In = typename decltype(in_pt)::type;
+            using Out = typename decltype(out_pt)::type;
+            using Accum = std::conditional_t<std::is_same<Out, Float16>{}, float, Out>;
+            struct Impl {
+                Accum Identity() { return Accum{1}; }
+                Accum MapIn(In in, int64_t /*index*/) { return static_cast<Accum>(in); }
+                void Reduce(Accum next, Accum& accum) { accum *= next; }
```

Comment:
(This is not relevant to the error)
It won't compile in gcc, complaining `*=` is invalid for `bool`s.

As `Accum` (and `Out`) is never actually `bool`, let's replace `VisitDtype` (the first one) at the line 98 with `VisitNumericDtype`, so that compilers won't try to generate the code for `bool`.

https://github.com/chainer/chainer/pull/6687/files#diff-5cc617697b5522640bcf9858d0c42b73R98

Reply:
Done.